core/server/master/src/main/java/alluxio/master/file/InodeTtlChecker.java

```diff
@@ -59,28 +61,42 @@ public InodeTtlChecker(FileSystemMaster fileSystemMaster, InodeTree inodeTree) {
 
   @Override
   public void heartbeat() throws InterruptedException {
-    Set<TtlBucket> expiredBuckets = mTtlBuckets.getExpiredBuckets(System.currentTimeMillis());
+    Set<TtlBucket> expiredBuckets = mTtlBuckets.pollExpiredBuckets(System.currentTimeMillis());
+    Map<Inode, Integer> failedInodesToRetryNum = new HashMap<>();
     for (TtlBucket bucket : expiredBuckets) {
-      for (Inode inode : bucket.getInodes()) {
+      for (Map.Entry<Long, Integer> inodeExpiryEntry : bucket.getInodeExpiries()) {
         // Throw if interrupted.
         if (Thread.interrupted()) {
           throw new InterruptedException("InodeTtlChecker interrupted.");
         }
+        long inodeId = inodeExpiryEntry.getKey();
+        int leftRetries = inodeExpiryEntry.getValue();
+        // Exhausted retry attempt to expire this inode, bail.
+        if (leftRetries <= 0) {
+          continue;
+        }
         AlluxioURI path = null;
         try (LockedInodePath inodePath =
             mInodeTree.lockFullInodePath(
-                inode.getId(), LockPattern.READ, NoopJournalContext.INSTANCE)
+                inodeId, LockPattern.READ, NoopJournalContext.INSTANCE)
         ) {
           path = inodePath.getUri();
         } catch (FileDoesNotExistException e) {
           // The inode has already been deleted, nothing needs to be done.
           continue;
         } catch (Exception e) {
-          LOG.error("Exception trying to clean up {} for ttl check: {}", inode.toString(),
-              e.toString());
+          LOG.error("Exception trying to clean up inode:{},path:{} for ttl check: {}", inodeId,
+              path, e.toString());
         }
         if (path != null) {
+          Inode inode = null;
           try {
+            inode = mTtlBuckets.loadInode(inodeId);
+            // Check again if this inode is indeed expired.
+            if (inode == null || inode.getTtl() == Constants.NO_TTL
+                || inode.getCreationTimeMs() + inode.getTtl() > System.currentTimeMillis()) {
+              continue;
```

Comment:
wanna remove this inode from the bucket (if not null)?

Reply:
check line 64 inside the "pollExpiredBuckets" function, it polled these buckets before processing